### ISSUE TITLE
Regression in 3.11.0: RequestInit Document not found

### ISSUE THREAD

**Issue by Nikerabbit** [facebook/hhvm#6648]
After upgrading from 3.10.1 to 3.11.0 and restarting hhvm, most of my MediaWiki stopped working and instead I get HTTP 404 with content "RequestInitDocument Not Found". Some pages still work, namely those not using index.php/SomethingHere syntax.

I am using nginx with [rewrite rules](https://github.com/wikimedia/translatewiki/blob/master/puppet/modules/nginx/files/translatewiki.net#L19).

**Comment by J0WI** [comment]
Do you use something like ```fastcgi_split_path_info ^(.+\.php)(/.+)$;```?

**Comment by nemobis** [comment]
> Do you use something like fastcgi_split_path_info ^(.+\.php)(/.+)$;?

I only see such a config for the mailman domain, not for the domain running MediaWiki:
> fastcgi_split_path_info (^/[^/]*)(.*)$;

https://phabricator.wikimedia.org/diffusion/GTWN/browse/master/?grep=fastcgi_split_path_info

**Comment by Nikerabbit** [comment]
@nemobis is correct, I don't.

My fastcgi.conf comes from nginx I believe:
``` 
fastcgi_param  SCRIPT_FILENAME    $document_root$fastcgi_script_name;
fastcgi_param  QUERY_STRING       $query_string;
fastcgi_param  REQUEST_METHOD     $request_method;
fastcgi_param  CONTENT_TYPE       $content_type;
fastcgi_param  CONTENT_LENGTH     $content_length;

fastcgi_param  SCRIPT_NAME        $fastcgi_script_name;
fastcgi_param  REQUEST_URI        $request_uri;
fastcgi_param  DOCUMENT_URI       $document_uri;
fastcgi_param  DOCUMENT_ROOT      $document_root;
fastcgi_param  SERVER_PROTOCOL    $server_protocol;
fastcgi_param  HTTPS              $https if_not_empty;

fastcgi_param  GATEWAY_INTERFACE  CGI/1.1;
fastcgi_param  SERVER_SOFTWARE    nginx/$nginx_version;

fastcgi_param  REMOTE_ADDR        $remote_addr;
fastcgi_param  REMOTE_PORT        $remote_port;
fastcgi_param  SERVER_ADDR        $server_addr;
fastcgi_param  SERVER_PORT        $server_port;
fastcgi_param  SERVER_NAME        $server_name;

# PHP only, required if PHP was built with --enable-force-cgi-redirect
fastcgi_param  REDIRECT_STATUS    200;
```

**Comment by lheredia** [comment]
I feel the same but using Apache, have any idea

**Comment by fredemmott** [comment]
Can you share your complete HHVM configuration files?

**Comment by lheredia** [comment]
Config file server.ini
; php options

pid = /var/run/hhvm/pid

; hhvm specific 

hhvm.server.port = 9000
hhvm.server.type = fastcgi
hhvm.server.default_document = index.php
hhvm.log.use_log_file = true
hhvm.log.file = /var/log/hhvm/error.log
hhvm.repo.central.path = /var/run/hhvm/hhvm.hhbc

Config file php.ini
; php options
session.save_handler = files
session.save_path = /var/lib/hhvm/sessions
session.gc_maxlifetime = 1440
upload_max_filesize = 200M
post_max_size = 200M

; hhvm specific 
hhvm.log.level = Warning
hhvm.log.always_log_unhandled_exceptions = true
hhvm.log.runtime_error_reporting_level = 8191
hhvm.mysql.typed_results = false
date.timezone = "America/Mexico_City"

**Comment by paulbiss** [comment]
There was only one fastcgi change in this release (c8340da976a4c939255fa845118d23c890ae7673). There were a few changes to the server directory that may have affected fastcgi transport though (https://github.com/facebook/hhvm/commits/HHVM-3.11/hphp/runtime/server). I don't have time to dig into them now but leaving this here to follow up on.

**Comment by Nikerabbit** [comment]
https://github.com/wikimedia/translatewiki/tree/master/puppet/modules/hhvm/files for my configs.

**Comment by beniboy87** [comment]
I'm also having this "RequestInitDocument Not Found" error on a fresh ubuntu server 14.04 install and nginx.

Nginx uses HHVM through unix socket, but I also tried with port 9000 - same response.

**Comment by J0WI** [comment]
Does anyone tried ```hhvm-repo-mode disable```? Repo mode can be broken after updating.

**Comment by nemobis** [comment]
J0WI, 16/12/2015 22:59:
> Does anyone tried |hhvm-repo-mode disable|? Repo mode can be broken
> after updating.

Did you find something in the configuration suggesting that it's enabled 
for translatewiki.net?

**Comment by J0WI** [comment]
This is not in the config, xou can enable it. See https://github.com/facebook/hhvm/wiki/Performance-Tuning#easy-way

**Comment by nemobis** [comment]
J0WI, 17/12/2015 00:06:
> This is not in the config, xou can enable it

It's not enabled in translatewiki.net.

**Comment by jurchiks** [comment]
@J0WI - thanks, `sudo hhvm-repo-mode enable /var/www/html` helped for me. I have a fresh setup of nginx and hhvm, and a normal PHP script set in that folder did not want to load, just kept spewing "RequestInit Document not found" and absolutely no info on why this was happening. Error logs contained only "Notice: File could not be loaded: /var/www/html/adminer.php", even though the file was clearly there.

**Comment by jurchiks** [comment]
Update: welp, that was a mistake. Executing that command made that one script work, but now I can't make another virtual host (root = /var/www/project/public) work, it throws the same error.
If I disable the repo-mode, then the virtual host works, but not the single script, and vice versa.
Why is it not possible to just have it serve any file anywhere?

**Comment by fredemmott** [comment]
Found via another issue, #6678

**Comment by fredemmott** [comment]
Those of you who reported this: can you reproduce it on the latest nightly? I think this is now fixed, but want some confirmation before putting into a 3.11.1.

**Comment by jurchiks** [comment]
From what I understood, my problem was for a particular script: https://www.adminer.org/static/download/4.2.3/adminer-4.2.3-mysql-en.php
However, that doesn't justify the fact that there was literally no information on why the file failed to load.

**Comment by fredemmott** [comment]
"Not Found" == there isn't a file where it's looking for one == it's looking for one in the wrong place.

Sorry, we need this to be tested by people that have the problem with 3.11; us trying to reproduce it and failing would not (and does not) confirm that the problem will be fixed for you.

**Comment by jurchiks** [comment]
Excuse me, but the error message says "Notice: File could not be loaded: /var/www/html/adminer.php", not "Not found". The file is there, it just refuses to load it and doesn't say why.

**Comment by fredemmott** [comment]
Ah, sorry, that's not the error message in the issue title; could still be a few different issues in this thread :(

**Comment by jurchiks** [comment]
Actually no, it is. The error message in the title is shown in the browser, while the "could not be loaded" message is in the hhvm.log.

**Comment by fredemmott** [comment]
Fair enough; still, are you able to test this with the hhvm-nightly packages?

**Comment by fredemmott** [comment]
^ with repo-mode disabled for a start. Best to make sure that works before re-enabling it.

**Comment by jurchiks** [comment]
Unless the nighly packages are available on Windows, then no. I was supposed to use hhvm at work for a project I was tasked with, but that is going down the drain. We use Linux exclusively at work.

**Comment by fredemmott** [comment]
Sorry, no, we have no support for Windows (including release versions) at the moment. Nightlies are available for Debian and Ubuntu.

**Comment by jurchiks** [comment]
Well in that case, you would have to wait at least 14 hours until I can (maybe) test it. Probably faster to do it yourself then...

**Comment by fredemmott** [comment]
I have tested one isolated repro, which no longer repros. Waiting is fine - even if we had the time to test more involved setups, it wouldn't actually confirm the issue is fixed for your environment.

Re-opening not because the issue isn't fixed, but just to get some more eyes on it and hopefully a confirmation.

**Comment by jurchiks** [comment]
Ok then, I'll try to test it tomorrow, if I get enough spare time at work.

**Comment by fredemmott** [comment]
Great, thanks :)

**Comment by jurchiks** [comment]
Welp, still the same error.
Need I share my nginx+hhvm configs?

**Comment by fredemmott** [comment]
:( as long as you're still getting RequestInitDocument not found somewhere, it's not the script. As many out of the following as possible would be useful:

 - nginx/hhvm/apache configs
 - being able to reproduce it with hhvm's built in HTTP server (or confirmation you can't)
 - being able to reproduce it when hhvm is running with HHVM config files explicitly specified as part of the command, or --no-config
 - being able to reproduce it with a standalone fastcgiclient (eg https://github.com/adoy/PHP-FastCGI-Client/) - then we don't have to worry about about apache/nginx configurations

**Comment by jurchiks** [comment]
nginx and hhvm configs: https://gist.github.com/jurchiks/7ead1fac58a55cec7654
nginx - default setup with the exception of the virtual hosts and installed hhvm fastcgi.
hhvm - set as replacement of `php` + configs and auto-start on boot.

Running `hhvm -m server -p 8080` while in `/var/www/html`, then accessing `http://localhost:8080/adminer.php` - it works. Damn. Well, that helps in localizing the problem, but I still don't know what causes it.

**Comment by fredemmott** [comment]
I think 0100248 fixed this; I'd appreciate it if anyone can test with tomorrow's nightlies when they're up.

**Comment by jurchiks** [comment]
Doesn't look like it did. Updated to latest nightly a few minutes ago, but http://localhost/adminer.php still gives me `RequestInitDocument Not Found`.

**Comment by fredemmott** [comment]
@jurchiks : are you using repo-authoritative mode? If so, try disabling it - and if that works, rebuild the repo. As far as we can tell, this issue is fixed.

We would appreciate reports from anyone else as to if their issue is fixed in the nightlies.

If anyone still has this problem with the nightlies with repo-authoritative mode disabled, to investigate we need at least one out of:

 - backtraces from all threads if it crashes(`thread apply all bt` in gdb)  - please use the nightly-dbg packages for this

 - an isolated way to reproduce the problem - ideally small PHP source, a command line for curl, and a command line for HHVM which starts with `hhvm --no-config` (specifying any required settings with `-d`)

An isolated repro case is much more useful than the backtraces, as then we're able to investigate more thoroughly and test to confirm the issue is fixed.

**Comment by jurchiks** [comment]
I'm not; like I wrote previously, I enabled it once to test if that helped, but it broke another project so I disabled it and haven't touched it since.

**Comment by tuminoid** [comment]
Just upgraded from Ubuntu 12.04 LTS to 14.04 LTS, and HHVM updated from 3.6.6 to 3.11.0, and I started getting this `RequestInitDocument Not Found` error for files that should be caught and rewritten by Nginx.

For example, `foo.php` doesn't exist and should be rewritten as `index.php?path=foo.php` as then processed by the app. Instead, hhvm throws above mentioned error. Redirecting `foo.bar` or any other non-HHVM file extension gets rewritten just fine. `.php` and `.hh` files error out. Nginx configuration didn't change and has worked fine for past year, with zero cases like this in the logs.

**Comment by fredemmott** [comment]
@tuminoid : it's fairly likely that this is fixed in the -nightly packages; I'd appreciate it if you could test them and confirm (several issues with this symptom have been fixed, just want some more confidence in how many before releasing 3.11.1)

**Comment by Nikerabbit** [comment]
`sudo dpkg -i hhvm-nightly_2016.01.11~trusty_amd64.deb`
`sudo service hhvm-development restart`

https://dev.translatewiki.net/wiki/Special:Version works and says `3.12.0-dev`

**Comment by fredemmott** [comment]
Thanks - that means it should be fixed for you in 3.11.1 and 3.12.0 :)

**Comment by fredemmott** [comment]
Closing as we now have 5 reports of the problem being fixed, and the issue @jurchiks is having is probably best investigated separately (and won't block a 3.11.1 - but if we find it and fix it in time, it's definitely a cherry-pick candidate).

@jurchiks: if that issue still exists for you, please open a separate issue (as it doesn't appear to be the same one that others are having, just a shared symptom), and:

 - try with a nightly-dbg package. This won't fix the problem, but might lead to an assertion failure instead, which might make debugging/isolating easier
 - provide an isolated repro that either uses `--no-config -d foo=bar` or `-c minimal_config_to_reproduce.ini` and either the built-in http server or the fastcgi-client linked above that lets us rule out nginx configuration
 - this error is supposed to mean that it can't open the file it's supposed to execute; problems in the PHP script itself should result in a 500, or a 404 Not Found that doesn't mention RequestInitDocument. Things this means to check include permissions on adminer.php, that re-creating the original index.php still works, and that `cat adminer.php > index.php` gives you (or doesn't give you) a working adminer on index.php, as another way to test for unusual permissions.

Please follow-up on a new issue if the problem still exists.
